crates/core/tedge_api/src/workflow/script.rs

```diff
@@ -0,0 +1,219 @@
+use crate::workflow::GenericStateUpdate;
+use crate::workflow::ScriptDefinitionError;
+use serde::de::Error;
+use serde::Deserialize;
+use serde::Deserializer;
+use serde::Serialize;
+use serde::Serializer;
+use serde_json::Value;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::os::unix::prelude::ExitStatusExt;
+
+/// A parsed Unix command line
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct ShellScript {
+    pub command: String,
+    pub args: Vec<String>,
+}
+
+/// Deserialize an Unix command line
+impl<'de> Deserialize<'de> for ShellScript {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let command_line = String::deserialize(deserializer)?;
+        let mut args = shell_words::split(&command_line)
+            .map_err(|err| D::Error::custom(format!("invalid script: {command_line}: {err}")))?;
+        if args.is_empty() {
+            Err(D::Error::custom("invalid script: empty"))
+        } else {
+            let script = args.remove(0);
+            Ok(ShellScript {
+                command: script,
+                args,
+            })
+        }
+    }
+}
+
+/// Serialize an Unix command line, using appropriate quotes
+impl Serialize for ShellScript {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+impl Display for ShellScript {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut args = vec![self.command.clone()];
+        args.append(&mut self.args.clone());
+        f.write_str(&shell_words::join(args))
+    }
+}
+
+/// Define how to interpret the exit code of a script as the next state for a command
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ExitHandlers {
+    on_success: Option<GenericStateUpdate>,
+    on_error: Option<GenericStateUpdate>,
+    on_kill: Option<GenericStateUpdate>,
+    on_exit: Vec<(u8, u8, GenericStateUpdate)>,
+    on_stdout: Vec<String>,
```

Comment:
> Are we making the user list out the expected state values here, just so that it is easier to figure out the entire control flow of a workflow, without having to audit the scripts? Or there are some other benefits as well?

Being able to compute the list of possible following states when in a given state will be a key to solve https://github.com/thin-edge/thin-edge.io/issues/2495.

> I see that they are unused now. Are we planning to validate that the output of the script is definitely one of the values provided in this list?

Indeed, not used for now. My former plan was to fail a command when moving to some unexpected state. But to solve the former issue we might have to be more subtle.